pyrdfstore/clean.py

```diff
@@ -0,0 +1,249 @@
+import logging
+import os
+import re
+from enum import Enum
+from functools import reduce
+from typing import Callable, List
+from urllib.parse import quote
+
+import validators
+from rdflib import BNode, Graph, Literal, URIRef
+
+log = logging.getLogger(__name__)
+
+
+class Level(Enum):
+    Graph = "graph"
+    Triple = "triple"
+    Node = "node"
+
+
+def reparse(g: Graph, format="nt"):
+    """This is a hack workaround for issue
+    https://github.com/RDFLib/rdflib/issues/2760
+    It reproduces the graph by serializing and parsing it again
+    Via an intermediate format (not jsonld!) that is known to work
+    :param g: the graph to reparse
+    :param format: the intermediate format to use
+    """
+    log.debug("reparse called")
+    return Graph().parse(data=g.serialize(format=format), format=format)
+
+
+reparse.level = Level.Graph
+
+
+def check_valid_uri(uri: str) -> bool:
+    """Checks if the uri is valid (does not contain chars it should not)
+    :param uri: the uri to check
+    :type uri: str
+    :return: True of uri is ok, else False"""
+    if uri.startswith("urn:"):
+        uri = "http://make.safe/" + uri
+    return bool(validators.url(uri))
+
+
+def clean_uri_str(uri: str, smart: bool = False) -> str:
+    """Escapes unacceptable chars in a URI.
+    :param smart: (optional) flag indicating smart-mode of operation.
+      Beinfg 'smart' indicates the routine will only clean if it is
+      needed. In other words: if True this does no cleaning is the
+      uri checks out to be already valid
+      - defaults to False, so executes a forced cleaning
+      Note that smart=True should make the result idempotent,
+      the non-smart-mode does not offer that guarantee.
+      Both guarantee a valid uri output though.
+    """
+    if smart and check_valid_uri(uri):
+        return uri
+    # else
+    return quote(uri, safe="~@#$&()*!+=:;,?/'")
+
+
+def clean_uri_node(ref: URIRef | BNode | Literal) -> URIRef | BNode | Literal:
+    """Escapes unacceptable chars in the URI of Graph Nodes
+    This by nature only affects Nodes of type URIRef.
+    """
+    log.debug("clean_uri_node called")
+    if not isinstance(ref, URIRef):
+        return ref  # nothing to do if not URIRef
+    # else
+    uri = str(ref)
+    if check_valid_uri(uri):
+        return ref  # nothing to do if uri is valid
+    # else
+    return URIRef(clean_uri_str(uri))
```

Comment:
How can smart cleaning be called is I wan to do this?

Reply:
you don´t, because you don 't need to

for cleaning nodes (unlike for cleaning uri) smartness is implied --> see test in 72
because of that we also don't pass smartness in line 75 and just force clean because we know it is needed already